### DISCUSSION TITLE
How to let an executable depends on a plain text file?

### DISCUSSION THREAD

**Discussion by Freed-Wu** [mesonbuild/meson#14056] [Q&A]
Some C compiler allow user to pass a plain text file, such as

`a.txt`:

```
# comment
-O2
--gcc
```

```sh
cc @a.txt main.c
```

If we simply

```meson
add_project_arguments('@' + 'a.txt')
executable('demo', ['main.c'])
```

When we update `a.txt`, the `demo` will not be rebuilt.

If we

```meson
add_project_arguments('@' + 'a.txt')
executable('demo', ['main.c', 'a.txt'])
```

```
ERROR: No specified compiler can handle file /the/path/of/a.txt
```

How can we do?

**Comment by bonzini** [comment]
You can use link_depends in some cases, but not for sources or arguments to the compiler (for example include paths). If you are interested in adding arguments to the compiler or the linker, you should instead [read the file](https://mesonbuild.com/Fs-module.html#read) and split it by word or line instead.

**Reply by Freed-Wu** [reply]
No, it is a compile depends. if we use link_depends, when a.txt is modified, it only relink demo, not recompile main.c.

**Reply by Freed-Wu** [reply]
So if I want to add arguments to compiler, I must read this file? For this case, I can do so. However, if is just `cc --option=XXX.txt`, how can I do? Can I have a `compile_depends` like `link_depends`?

**Reply by eli-schwartz** [reply]
When including args via `@file` you can let ninja automatically create response files for you and include the arguments directly, as suggested above.

For compiler options that process the contents of the file as compiled data this will be handled by the compiler itself via depfiles.

**Reply by eli-schwartz** [reply]
Such a Makefile would be weird because it doesn't actually depend on or use dep_file.txt and therefore it should not be rebuilt after all.

Also, your example Makefile shows dep_file.txt as a link-time dependency

**Reply by Freed-Wu** [reply]
If an executable depends on a file, when the file is modified (time stamp is changed), this executable must be rebuilt, In Makefile, It looks like:

```make
main$(EXE): main.o
  cc $< -o $@
main.o: main.c dep_file.txt
  cc --some-option=dep_file.txt $< -c -o $@
```

Now How can I do same thing in meson?

**Reply by eli-schwartz** [reply]
Here is a worked Makefile example:

```make
main$(EXE): main.o
	cc $< -o $@
main.o: main.c
	cc -include dep_file.txt $< -c -o $@ -MD -MF $@.d

-include *.d
```

Meson automatically handles -MD -MF for you, and automatically includes `*.d` as well. And `main.o` will always be rebuilt every time dep_file.txt is modified, due to those *.d files.

Please provide an example project where this does not work (one where you pass `dep_file.txt` as an argument which is NOT `@dep_file.txt`, but does not get rebuilt).

Avoid ambiguous examples such as `--option=XXX.txt`:
```
$ gcc --option=XXX.txt
gcc: error: unrecognized command-line option ‘--option=XXX.txt’
```

**Reply by bonzini** [reply]
Can you provide an actual example of what --some-option would be?

**Reply by Freed-Wu** [reply]
In my situation, C compiler is TI compiler, which use `--cmd_file=XXX.cmd`.
